server/src/project_service.ts

```diff
@@ -0,0 +1,142 @@
+import * as ts from 'typescript/lib/tsserverlibrary'; // used as value
+import * as lsp from 'vscode-languageserver';
+import {tsDiagnosticToLspDiagnostic} from './diagnostic';
+import {filePathToUri} from './utils';
+
+// NOTE:
+// There are three types of `project`:
+// 1. Configured project - basically all source files that belong to a tsconfig
+// 2. Inferred project - other files that do not belong to a tsconfig
+// 3. External project - not used in this context
+// For more info, see link below.
+// https://github.com/Microsoft/TypeScript/wiki/Standalone-Server-%28tsserver%29#project-system
+
+/**
+ * `ProjectService` is a singleton service for the entire lifespan of the
+ * language server. This specific implementation is a very thin wrapper
+ * around TypeScript's `ProjectService`. On creation, it spins up tsserver and
+ * loads `@angular/language-service` as a tsserver plugin.
+ * Care must be taken to load `@angular/language-service` from the project or
+ * workspace. This extension shoud not bundle its own
+ * `@angular/language-service`, not even as a backup. This is to ensure that
+ * the extension becomes a no-op when non-Angular TS projects are opened.
```

Comment:
I thought in the case that an only-TS project is opened, the language service still runs but _always_ defers to the TypeScript language service. Is this not correct?

Can we guarantee that Angular users will have their own `@angular/language-service`? Do they have to specify it as a dependency for the plugin to work?

Reply:
The first statement is incorrect. The goal of the Angular LS is to answer Angular queries only. That's why the `angularOnly` flag was introduced in the plugin.

Our extension loads `@angular/language-service` as a global plugin.

Users who want to use the Angular LS must install `@angular/language-service` in their project (it's installed by default if users use the Angular CLI). If the extension does not find the package in the workspace, it should bail, effectively becoming a no-op extension.

That said, if `@angular/language-service` is loaded as a **tsserver local plugin**, it should never disable TypeScript language service. Does that make sense?

Reply:
When would the Angular language service be loaded as a global plugin versus a local plugin? If it is a global plugin and angularOnly, would this mean that TypeScript language service would not be enabled?

You’ve explained this to me before, but I think my mental model has degraded; sorry about that :)

Reply:
Instead of explaining the difference between local plugin and global plugin here, I'll write a doc in the wiki and capture the previous discussion. I started a page already, will add more entries to it later.